### PR TITLE
Add JUnit extension for attaching log record listener to test cases

### LINKED ISSUE #2546: Reduce duplicated code to test log records
### 🤔 What's the problem you've observed?

We use a `LogRecordListener` to capture log messages and test them. This results in a bit of boiler plate code repeated over several tests:

https://github.com/cucumber/cucumber-jvm/blob/78adb8fe2e4c9d80ccfc36b1a69a8a27bab4ace7/core/src/test/java/io/cucumber/core/runtime/FeaturePathFeatureSupplierTest.java#L30-L39

### ✨ Do you have a proposal for making it better?

Implement [a JUnit 5 extension](https://junit.org/junit5/docs/current/user-guide/#extensions) that provides the `LogRecordListener` as a method argument to tests.

```java
@WithLogRecordListener
class Example {

    @Test
    void example(LogRecordListener logRecordListener){

    }
}
```    

Then replace the existing usages with this extension.

### :hammer_and_wrench: Implementation hints

JUnit 5 supports meta-annotations. So something like this can make it possible in a single annotation:

```java
@Target({ ElementType.TYPE, ElementType.METHOD })
@Retention(RetentionPolicy.RUNTIME)
@ExtendWith(WithLogRecordListener.Extension.class)
public @interface WithLogRecordListener {

	class Extension implements BeforeEachCallback, AfterEachCallback, ParameterResolver {
        // Implement methods 
	}
}
```

#### :test_tube: Testing hints

Since we're re-factoring the existing tests I don't expect any additional tests for the extension itself.

### PR DESCRIPTION
**Pull request by domcorvasce**
You can use the `@WithLogRecordListener` annotation to extend your test class.

```java
    @WithLogRecordListener
    class FooTest {
```

It will attach an instance of `LogRecordListener` to each test method.
The method can access the class instance by accepting an argument which is automatically resolved by the extension:

```java
    @Test
    void fooBar(LogRecordListener logRecordListener) {
        LogRecord logRecord = logRecordListener.getLogRecords().get(0);
```

### ⚡️ What's your motivation? 

Fixes #2546 

### 🏷️ What kind of change is this?

- :bank: Refactoring/debt/DX (improvement to code design, tooling, documentation etc. without changing behaviour)

### ♻️ Anything particular you want feedback on?

--

### 📋 Checklist:

- [x] I agree to respect and uphold the [Cucumber Community Code of Conduct](https://cucumber.io/conduct/)
- [ ] I've changed the behaviour of the code
  - [ ] I have added/updated tests to cover my changes.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.
- [ ] Users should know about my change
  - [ ] I have added an entry to the "Unreleased" section of the [**CHANGELOG**](../CHANGELOG.md), linking to this pull request.

----

*This text was originally generated from a [template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/about-issue-and-pull-request-templates), then edited by hand. [You can modify the template here.](https://github.com/cucumber/.github/edit/main/.github/PULL_REQUEST_TEMPLATE.md)*

### CODE REVIEW

File: core/src/test/java/io/cucumber/core/logging/WithLogRecordListener.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+        @Override
+        public void afterEach(ExtensionContext context) throws Exception {
+            LoggerFactory.removeListener(logRecordListener);
+        }
+
+        @Override
+        public boolean supportsParameter(ParameterContext paramContext, ExtensionContext extContext)
+                throws ParameterResolutionException {
+            return paramContext.getParameter().getType() == LogRecordListener.class;
```
Comment (mpkorstanje) [inline core/src/test/java/io/cucumber/core/logging/WithLogRecordListener.java:39]: Parameters can be provided to both constrictors and methods. However log record listeners can only be used in method scoped.

File: core/src/test/java/io/cucumber/core/logging/WithLogRecordListener.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target({ ElementType.TYPE, ElementType.METHOD })
+@Retention(RetentionPolicy.RUNTIME)
+@ExtendWith({ WithLogRecordListener.Extension.class })
+public @interface WithLogRecordListener {
+    class Extension implements BeforeEachCallback, AfterEachCallback, ParameterResolver {
+        protected LogRecordListener logRecordListener;
```
Comment (mpkorstanje) [inline core/src/test/java/io/cucumber/core/logging/WithLogRecordListener.java:23]: Because tests maybe run in parallel with JUnit, it's generally good to keep extensions stateless and store any state in the context.

File: junit-platform-engine/src/test/java/io/cucumber/junit/platform/engine/DiscoverySelectorResolverTest.java [hunk 3]
```diff
@@ -51,25 +52,19 @@
 import static org.junit.platform.engine.discovery.DiscoverySelectors.selectUniqueId;
 import static org.junit.platform.engine.discovery.DiscoverySelectors.selectUri;
 
+@WithLogRecordListener
```
Comment (mpkorstanje) [inline junit-platform-engine/src/test/java/io/cucumber/junit/platform/engine/DiscoverySelectorResolverTest.java:55]: Unfortunately due to the architecture we can't share test code between modules in this project. I think that duplicating the class would not be the least disturbing solution. While it duplicated the code, it preserves the modular nature of the project.

**Review by mpkorstanje — COMMENTED** [review]
Thanks for the PR! At a glance I did see a problem. But I'm currently taking a break. Will have a better look when back.

**Comment by domcorvasce** [comment]
@mpkorstanje hey there, I did push the changes as requested.
Thanks for the patience, it's my first time playing around with JUnit 5 extensions 😅

File: core/src/test/java/io/cucumber/core/logging/WithLogRecordListener.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            // If the test class is not available, fall back to the GLOBAL
+            // namespace
+            if (testClass.isPresent()) {
+                namespaceId = testClass.get().getSimpleName();
+            }
+
+            ExtensionContext.Namespace namespace = ExtensionContext.Namespace.create(namespaceId);
+            ExtensionContext.Store store = context.getStore(namespace);
+
+            return store;
```
Comment (mpkorstanje) [inline core/src/test/java/io/cucumber/core/logging/WithLogRecordListener.java:38]: Try name spacing the store to the class of this extension and the test method. That should be sufficiently unique. And because you've already checked there is a test method in `supportsParameter` you won't have to check it again here. You can safely assume it to be present.

File: core/src/test/java/io/cucumber/core/logging/WithLogRecordListener.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+        @Override
+        public void beforeEach(ExtensionContext context) throws Exception {
+            LogRecordListener logRecordListener = new LogRecordListener();
+            LoggerFactory.addListener(logRecordListener);
+
+            // Store the log record listener instance into a namespaced store
+            // to prevent issues when running tests in parallel
+            ExtensionContext.Store store = getContextStore(context);
+            store.put("logRecordListener", logRecordListener);
```
Comment (mpkorstanje) [inline core/src/test/java/io/cucumber/core/logging/WithLogRecordListener.java:49]: There is a trick you can use here, it is called lazy initiation. Instead of creating the object and storing it, you can check the store and create if not present. Have a look at `store.getOrComputeIfAbsent`.

File: core/src/test/java/io/cucumber/core/logging/WithLogRecordListener.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        }
+
+        @Override
+        public void afterEach(ExtensionContext context) throws Exception {
+            ExtensionContext.Store store = getContextStore(context);
+            LoggerFactory.removeListener((LogRecordListener) store.get("logRecordListener"));
+            store.remove("logRecordListener");
```
Comment (mpkorstanje) [inline core/src/test/java/io/cucumber/core/logging/WithLogRecordListener.java:56]: The store is cleaned after test. No need for cleanup!

**Review by mpkorstanje — COMMENTED** [review]
Getting there! There are a few tricks you can use to remove edge cases and make the code simpler.

Still away btw. Dunno when I'll be able to review again.

**Comment by aslakhellesoy** [comment]
Hi @domcorvasce,

Thanks for your making your first contribution to Cucumber, and welcome to the Cucumber committers team! You can now push directly to this repo and all other repos under the cucumber organization! 🍾

In return for this generous offer we hope you will:

* ✅ Continue to use branches and pull requests. When someone on the core team approves a pull request (yours or someone else's), you're welcome to merge it yourself.
* 💚 Commit to setting a good example by following and upholding our [code of conduct](https://cucumber.io/conduct) in your interactions with other collaborators and users.
* 💬 Join the [community Slack channel](http://cucumber.io/support#slack) to meet the rest of the team and make yourself at home.
* ℹ️ Don't feel obliged to help, just do what you can if you have the time and the energy.
* 🙋 Ask if you need anything. We're looking for feedback about how to make the project more welcoming, so please tell us!

On behalf of the Cucumber core team,
Aslak Hellesøy
Creator of Cucumber

**Comment by mpkorstanje** [comment]
Thanks a bucket!. Did a bit of clean and merged the changes.
